http-server/src/middlewares/proxy_request.rs

```diff
@@ -0,0 +1,174 @@
+//! Middleware that proxies requests at a specified URI to internal
+//! RPC method calls.
+
+use crate::response;
+use futures_util::ready;
+use hyper::body::HttpBody;
+use hyper::header::{ACCEPT, CONTENT_TYPE};
+use hyper::http::HeaderValue;
+use hyper::{Body, Method, Request, Response};
+use pin_project_lite::pin_project;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tower::{Layer, Service};
+
+/// Layer that applies [`ProxyRequest`] which proxies the `GET /path` requests to
+/// specific RPC method calls and that strips the response.
+///
+/// See [`ProxyRequest`] for more details.
+#[derive(Debug, Clone)]
+pub struct ProxyRequestLayer {
+	path: String,
+	method: String,
+}
+
+impl ProxyRequestLayer {
+	/// Creates a new [`ProxyRequestLayer`].
+	///
+	/// See [`ProxyRequest`] for more details.
+	pub fn new(path: impl Into<String>, method: impl Into<String>) -> Self {
+		Self { path: path.into(), method: method.into() }
+	}
+}
+impl<S> Layer<S> for ProxyRequestLayer {
+	type Service = ProxyRequest<S>;
+
+	fn layer(&self, inner: S) -> Self::Service {
+		ProxyRequest::new(inner, self.path.clone(), self.method.clone())
+	}
+}
+
+/// Proxy `GET/path` requests to the specified RPC method calls.
+///
+/// # Request
+///
+/// The `GET /path` requests are modified into valid `POST` requests for
+/// calling the RPC method. This middleware adds appropriate headers to the
+/// request, and completely modifies the request `BODY`.
+///
+/// # Response
+///
+/// The response of the RPC method is stripped down to contain only the method's
+/// response, removing any RPC 2.0 spec logic regarding the response' body.
+#[derive(Debug, Clone)]
+pub struct ProxyRequest<S> {
+	inner: S,
+	path: String,
+	method: String,
```

Comment:
Perhaps `Arc<str>` would be better since you clone them per request and don't need to modify them :)

Reply:
It makes sense, thanks for the optimization feedback 😄 !